sql/sql_acl.cc

```diff
@@ -3338,10 +3338,14 @@ static int check_user_can_set_role(THD *thd, const char *user,
                                                 check_role_is_granted_callback,
                                                 NULL) == -1))
       {
-        /* Role is not granted but current user can see the role */
-        my_printf_error(ER_INVALID_ROLE, "User %`s@%`s has not been granted role %`s",
-                        MYF(0), thd->security_ctx->priv_user,
-                        thd->security_ctx->priv_host, rolename);
+        /* if --skip-name-resolve is used*/
+        if (!host)
+          my_printf_error(ER_INVALID_ROLE, "User %`s@%`s has not been granted role %`s",
+                          MYF(0), thd->security_ctx->priv_user, thd->security_ctx->priv_host, rolename);
+        else
+          /* Role is not granted but current user can see the role */
+          my_printf_error(ER_INVALID_ROLE, "User %`s@%`s has not been granted role %`s",
+                          MYF(0), user, host, rolename);
```

Comment:
Ok, this is not trivial.

On the first glance this is very wrong.
The situation here: `priv_user@priv_host` tries to set the role `rolename` for the account `user@host`.
The correct error should obviously print `user@host` — that's whom the role cannot be granted to.
The fact that under `--skip-name-resolve` the patch for some reason prints the current user, not the target user, looks like a bug.

But `check_user_can_set_role()` is used in two cases. For `SET DEFAULT ROLE xxx TO yyy`, in that case `host` is never NULL, so it prints the correct error message. And for `SET ROLE xxx`, in that case the current user is always the target user, `host` can be NULL, but printing `priv_user@priv_host` in the error message is not incorrect, it's just confusing.

So, two options here: either add the above explanation here, to clarify why using `priv_host@priv_user` is not a bug in this particular case. Or remove the `if` and use
```c
        /* Role is not granted but current user can see the role */
        my_printf_error(ER_INVALID_ROLE, "User %`s@%`s has not been granted role %`s",
                        MYF(0), user, host ? host : "%", rolename);
```